front/src/applications/stdcm/components/StdcmLayersModal.tsx

```diff
@@ -0,0 +1,157 @@
+      { infraId: infraID! },
+      { skip: !infraID }
+    );
+  const { data: speedLimitTags } = osrdEditoastApi.endpoints.getSpeedLimitTags.useQuery();
+
+  const allSpeedLimitTags = compact(concat(speedLimitTags, speedLimitTagsByInfraId));
+  const allSpeedLimitTagsOrdered = useMemo(() => allSpeedLimitTags.sort(), [allSpeedLimitTags]);
+
+  const DEFAULT_SPEED_LIMIT_TAG = useMemo(() => t('Editor.layers-modal.noSpeedLimitByTag'), [t]);
+
+  const speedLimitOptions = useMemo(
+    () => uniq([DEFAULT_SPEED_LIMIT_TAG, ...allSpeedLimitTagsOrdered]),
+    [allSpeedLimitTagsOrdered, selectedLayers]
+  );
+  const toggleLayer = useCallback(
+    (layer: keyof LayersSettings) => {
+      const isEnabled = !selectedLayers[layer];
+
+      const updatedLayers = {
+        ...layersSettings,
+        [layer]: isEnabled,
+        // Preserve speedlimittag unless the layer is disabled
+        speedlimittag:
+          layer === ('speedlimittag' as keyof LayersSettings) && !isEnabled
+            ? null
+            : layersSettings.speedlimittag,
+      };
+
+      setSelectedLayers(updatedLayers);
+      onChange({ newLayers: updatedLayers });
+    },
+    [dispatch, selectedLayers, onChange, layersSettings]
+  );
+
+  return (
+    <Modal title={t('Editor.nav.toggle-layers')}>
+      <div className="container-fluid mb-3">
+        <div>
+          <h4>{t('Editor.nav.osrd-layers')}</h4>
+        </div>
+        <div className="row">
+          {LAYERS.map(({ layer, icon }) => (
+            <div className="col-lg-6" key={`${layer}`}>
+              <div className="d-flex align-items-center mt-2">
+                <SwitchSNCF
+                  type="switch"
+                  checked={!!selectedLayers[layer as keyof LayersSettings]}
+                  onChange={() => toggleLayer(layer as keyof LayersSettings)}
+                  name={`editor-layer-${layer}`}
+                  id={`editor-layer-${layer}`}
+                />
+                {isString(icon) ? (
+                  <img className="layer-modal-img mx-2" src={icon} alt="" />
+                ) : (
+                  <div>{icon}</div>
+                )}
+                <div className="d-flex flex-column">
+                  <div>{t(`Editor.layers.${layer}`)}</div>
+                </div>
+              </div>
+            </div>
+          ))}
+        </div>
+        <hr />
+        <div>
+          <h4>{t('Editor.nav.speed-limits')}</h4>
+          <select
+            id="speedLimitTag"
+            className="form-control"
+            disabled={!selectedLayers.speed_limits}
+            value={layersSettings.speedlimittag || DEFAULT_SPEED_LIMIT_TAG}
+            onChange={(e) => {
+              const newTag = e.target.value !== DEFAULT_SPEED_LIMIT_TAG ? e.target.value : null;
+              dispatch(
+                updateStdcmLayers({
+                  ...layersSettings,
+                  speedlimittag: newTag,
+                })
+              );
```

Comment:
Before dispatching the layers here, you should update the layersSettings here in the component ;)
```suggestion
              const newLayers = { ...layersSettings, speedlimittag: newTag };
              setLayersSettings(newLayers);
              onChange(newLayers);
```